lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -530,6 +533,88 @@ def internal_check_future_reserved_word(o, name)
     end
   end
 
+  def internal_check_file_scope(o, name, file)
+    return if file.nil?
+
+    file_scope = @file_scope_hash[file]
+    if file_scope.nil?
+      return if @file_scope_hash.key?(file)
+      file_scope = @file_scope_hash[file] = generate_file_scope file
+      return if file_scope.nil?
+    end
+
+    if check_scope(name, file_scope)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    end
+    #Return to direct call after benchmarking check method
+    #if name.start_with? file_scope
+    #if name !~ file_scope
+    #  acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    #end
+  end
+
+  def create_scope_match(names)
+    qualified_name = names.join("::")
+
+    #qualified_name
+    Regexp.new "^#{qualified_name}", Regexp::IGNORECASE
+  end
+
+  def check_scope(name, file_scope)
+    #!name.start_with?(file_scope)
+    name !~ file_scope
+  end
+
+  def generate_file_scope(file)
+    path = Pathname.new(file)
+
+    return nil if path.extname != ".pp"
+
+    return nil if initial_manifest(file)
+
+    path = Pathname.new(file).each_filename.to_a
+
+    definition_dir_index = find_module_definition_dir path
+
+    # How can we get this result?
+    # If it is not an initial manifest, it must come from a module,
+    # and from the manifests dir there.  This may never get used...
+    return NEVER_MATCH if definition_dir_index.nil?
+
+    before_definition_dir = definition_dir_index == 0 ? 0 : definition_dir_index - 1
+    after_definition_dir = definition_dir_index + 1
+    names = path[after_definition_dir .. -2] # Directories inside module
+    names.unshift(path[before_definition_dir]) # Name of the module itself
+    # Do not include name of module init file at top level of module
+    if !(path.length == (after_definition_dir+1) && path[-1] == 'init.pp')
+      names.push(path[-1][0 .. -4]) # Remove .pp from filename
+    end
+
+    create_scope_match names
```

Comment:
can be changed to `names.join('::').freeze` since start_with? is faster than regexp